### PR TITLE
Turbopack build: Fix middleware matchers implementation

### PR DESCRIPTION
**Pull request by timneutkens**
Found that middleware matchers were not fully implemented. Copied the implementation we have for webpack as much as possible. Had to add a workaround for path-to-regexp being used so now we post-processes the manifest that the compiler outputs to turn the path-to-regexp string to a regex.

### CODE REVIEW

**Comment by timneutkens** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/vercel/next.js/69602?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#69602** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69602?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#69534** <a href="https://app.graphite.dev/github/pr/vercel/next.js/69534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @timneutkens and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
## Failing test suites 
Commit: e155cc9a9bb8d3cd4a0dd64cbcaac189936523fd

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/app-dir/logging/fetch-logging.test.ts` (PPR)
- app-dir - logging > with fetches verbose logging > when logging.fetches.hmrRefreshes is true > should log requests for HMR refreshes

<details>
<summary>Expand output</summary>

  ● app-dir - logging › with fetches verbose logging › when logging.fetches.hmrRefreshes is true › should log requests for HMR refreshes

    expect(received).toBe(expected) // Object.is equality

    Expected: "Hello Test!"
    Received: "Hello World!"

      276 |                 return retry(async () => {
      277 |                   headline = await browser.waitForElementByCss('h1').text()
    > 278 |                   expect(headline).toBe('Hello Test!')
          |                                    ^
      279 |
      280 |                   const logs = stripAnsi(
      281 |                     next.cliOutput.slice(outputIndex)

      at toBe (e2e/app-dir/logging/fetch-logging.test.ts:278:36)
      at retry (lib/next-test-utils.ts:806:14)
      at lib/next-modes/next-dev.ts:184:9
      at NextDevInstance.patchFile (lib/next-modes/base.ts:512:9)
      at Object.<anonymous> (e2e/app-dir/logging/fetch-logging.test.ts:268:13)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: test/e2e/middleware-trailing-slash/test/index.test.ts [hunk 1]
```diff
@@ -114,15 +114,17 @@ describe('Middleware Runtime trailing slash', () => {
         }
         delete middlewareWithoutEnvs.env
         expect(middlewareWithoutEnvs).toEqual({
-          files: expect.arrayContaining([
-            'server/edge-runtime-webpack.js',
-            'server/middleware.js',
-          ]),
+          files: process.env.TURBOPACK
+            ? expect.toBeArray()
```
Comment (ForsakenHarmony) [inline test/e2e/middleware-trailing-slash/test/index.test.ts:118]: Should we at least assert a length here?

File: test/e2e/middleware-trailing-slash/test/index.test.ts [hunk 2]
```diff
@@ -114,15 +114,17 @@ describe('Middleware Runtime trailing slash', () => {
...
+            : expect.arrayContaining([
+                'server/edge-runtime-webpack.js',
+                'server/middleware.js',
+              ]),
           name: 'middleware',
           page: '/',
           matchers: [{ regexp: '^/.*$', originalSource: '/:path*' }],
           wasm: [],
-          assets: [],
+          assets: process.env.TURBOPACK ? expect.toBeArray() : [],
```
Comment (ForsakenHarmony) [inline test/e2e/middleware-trailing-slash/test/index.test.ts:127]: Maybe here as well?

File: test/e2e/middleware-trailing-slash/test/index.test.ts [hunk 3]
```diff
@@ -132,9 +134,11 @@ describe('Middleware Runtime trailing slash', () => {
         )
         for (const key of Object.keys(manifest.middleware)) {
           const middleware = manifest.middleware[key]
-          expect(middleware.files).toContainEqual(
-            expect.stringContaining('server/edge-runtime-webpack')
-          )
+          if (!process.env.TURBOPACK) {
```
Comment (ForsakenHarmony) [inline test/e2e/middleware-trailing-slash/test/index.test.ts:137]: And here?

File: crates/next-api/src/middleware.rs [hunk 4]
```diff
@@ -156,15 +156,64 @@ impl MiddlewareEndpoint {
...
+                    // Mirrors implementation in get-page-static-info.ts getMiddlewareMatchers
+                    let mut source = matcher.original_source.to_string();
+                    let is_root = source == "/";
+                    let has_locale = matcher.locale;
+
+                    if has_i18n_locales && has_locale {
+                        source = format!(
+                            "/:nextInternalLocale((?!_next/)[^/.]{{1,}}){}",
+                            if is_root {
+                                "".to_string()
```
Comment (ForsakenHarmony) [inline crates/next-api/src/middleware.rs:189]: This indentation is getting a bit much, can we reduce it somehow?

**Review by ForsakenHarmony — CHANGES_REQUESTED** [review]

File: test/e2e/middleware-trailing-slash/test/index.test.ts (hunk 1 above)
Comment (wbinnssmith) [inline test/e2e/middleware-trailing-slash/test/index.test.ts:118]: is this now too dynamic?

**Review by wbinnssmith — APPROVED** [review]

File: test/e2e/middleware-trailing-slash/test/index.test.ts (hunk 1 above)
Comment (timneutkens) [reply test/e2e/middleware-trailing-slash/test/index.test.ts:118]: The length can change at any point when the chunking changes and it's not really relevant, we already check if the edge function can run and return the right result.

Comment (timneutkens) [reply test/e2e/middleware-trailing-slash/test/index.test.ts:118]: Yeah: https://github.com/vercel/next.js/pull/69602#discussion_r1742383593

Comment (timneutkens) [reply test/e2e/middleware-trailing-slash/test/index.test.ts:118]: Or better said: the webpack one is not dynamic enough 😄

Comment (ForsakenHarmony) [reply test/e2e/middleware-trailing-slash/test/index.test.ts:118]: could assert >0

**Pull request merged**